### PR TITLE
Proxy support for websocket clients.

### PR DESCRIPTION
**Pull request by wolfeidau**
Firstly the intent of this pull request is to enable websocket clients to traverse proxy servers using the standard environment variables in line with the golang http package.

The reason I am submitting this patch is one of the upstream consumers of this library currently doesn't work when run within an isolated network environment (see https://github.com/aws/amazon-ecs-agent/issues/23).

I have tried to incorporate a simple and easily overridable API along with test coverage for my additions.

### CODE REVIEW

**Comment by garyburd** [comment]
- Structure the code so the [net/http ProxyFromEnvironment](https://godoc.org/net/http#ProxyFromEnvironment) function can be used as is. 

- Specify proxy function using Dialer field instead of package level variable.  As with net/http, the default is no proxy.

        // Proxy specifies a function to return a proxy for a given
        // Request. If the function returns a non-nil error, the
        // request is aborted with the provided error.
        // If Proxy is nil or returns a nil *URL, no proxy is used.
        Proxy func(*http.Request) (*url.URL, error)

- I will submit a PR in a few hours to move the code from NewClientConn to Dialer.Dial and rewrite NewClientConn using Dialer.Dial.  With this change, the proxy code can use the  connection's [buffered reader](https://github.com/gorilla/websocket/blob/423912737d2c26382ab4b2647d796811d4efa00b/conn.go#L161) to read the response from the proxy server.

**Comment by garyburd** [comment]
See refactored code at https://github.com/garyburd/websocket/commit/ae2bbac1c1729aea27efd3cb2c168c08becf572a.  I want to add some tests before pushing it to master.  

The the net/http DefaultTransport uses ProxyFromEnvironment.  The websocket DefaultClient should as well.

**Comment by wolfeidau** [comment]
Just a note you don't want to use a HTTP proxy for websockets as it is just a primitive request relay, websockets will need to always use the HTTPS proxy with the CONNECT method.

This is one of the reasons I forked and adjusted the `ProxyFromEnvironment`.

That said more alignment between would be idea!

**Comment by garyburd** [comment]
What's the reason for not using the CONNECT method on an HTTP_PROXY?

**Comment by wolfeidau** [comment]
Good point they will both typically be the same these days, although a lot of the older proxies still have ACLs for either HTTP or HTTPS both are always permitted.

**Comment by garyburd** [comment]
The main issue is deciding on the protocol to use when calling the `Proxy func(*http.Request) (*url.URL, error)` function. Should the package pass the handshake request as is to the function or should the package always set the protocol to "https"?   

Hopefully we don't need the forked code. I don't want to figure how to bring code in the project with a different license.

**Comment by wolfeidau** [comment]
I absolutely agree on the licence side, I can understand why your loath to add the code as it is, that is why I made the intent of the change very clear.

I would recommend forcing use of the HTTPS proxy just to be more "correct", if that requires tweaking the URL to convince that method to give you the URL of HTTPS Proxy I think that is OK. 

Just my two cents.

**Comment by garyburd** [comment]
The Dial refactoring is pushed to master.  Please do the following:

- Add Proxy field to Dialer.

        // Proxy specifies a function to return a proxy for a given
        // Request. If the function returns a non-nil error, the
        // request is aborted with the provided error.
        // If Proxy is nil or returns a nil *URL, no proxy is used.
        Proxy func(*http.Request) (*url.URL, error)

- Set DefaultDialer.Proxy to http.ProxyFromEnvironment.

- Write the test following the structure of TestHostHeader.  Instead of defining another test server type, [hook the handler in the existing test server](https://github.com/gorilla/websocket/blob/5ed2f4547d7b9c036541b27a67ecf54e41756997/client_server_test.go#L299-L306).

File: client.go [hunk 1]
```diff
@@ -220,6 +274,32 @@ func (d *Dialer) Dial(urlStr string, requestHeader http.Header) (*Conn, *http.Re
...
+			Host:   hostPort,
+			Header: make(http.Header),
+		}
+
+		connectReq.Write(netConn)
+
+		// Read response.
+		// Okay to use and discard buffered reader here, because
+		// TLS server will not speak until spoken to.
+		br := bufio.NewReader(netConn)
```
Comment (garyburd) [inline client.go]: Use c.br. No need to create another one.

File: client.go [hunk 2]
```diff
@@ -220,6 +274,32 @@ func (d *Dialer) Dial(urlStr string, requestHeader http.Header) (*Conn, *http.Re
...
+		resp, err := http.ReadResponse(br, connectReq)
+		if err != nil {
+			netConn.Close()
```
Comment (garyburd) [inline client.go]: Deferred function will close netConn. Don't close it here.

File: client.go [hunk 3]
```diff
@@ -220,6 +274,32 @@ func (d *Dialer) Dial(urlStr string, requestHeader http.Header) (*Conn, *http.Re
...
+			return nil, nil, err
+		}
+		if resp.StatusCode != 200 {
+			f := strings.SplitN(resp.Status, " ", 2)
+			netConn.Close()
```
Comment (garyburd) [inline client.go]: Deferred function will close netConn. Don't close it here.

File: client_server_test.go [hunk 4]
```diff
@@ -123,6 +123,47 @@ func sendRecv(t *testing.T, ws *Conn) {
...
 }
 
+func TestProxyDial(t *testing.T) {
+
+	s := newServer(t)
+	defer s.Close()
+
+	surl, _ := url.Parse(s.URL)
+
+	cstDialer.Proxy = func(r *http.Request) (*url.URL, error) {
```
Comment (garyburd) [inline client_server_test.go]: cstDialer.Proxy = http.ProxyURL(surl)

File: client.go (hunk 1 above)
Comment (wolfeidau) [inline client.go]: I had a look at doing this but the `conn` isn't created till after the TLS handshake, which matches the core API, hence the need for the `br` at this stage.

**Comment by wolfeidau** [comment]
Latest round of fixes is up and tested with the caveat / question / comment above.

Cheers

File: client.go (hunk 1 above)
Comment (garyburd) [inline client.go]: Ah, got it.

**Comment by garyburd** [comment]
I was informed that client.go contains copyrighted code. I need to figure out what to do with the licensing.

**Comment by wolfeidau** [comment]
I can rewrite the helper methods I have used from the golang API if required, this use case is bit more specific anyway.

**Comment by garyburd** [comment]
Replace use of canonicalAddr and supporting functions with existing function hostPortNoPort.  Update hostPortNoPort to handle "http" and "https".

**Comment by wolfeidau** [comment]
Sterling idea, done, tested and pushed.

**Pull request merged**